src/linear_algebra/orientation.lean

```diff
@@ -0,0 +1,376 @@
+/-
+Copyright (c) 2021 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import linear_algebra.determinant
+
+/-!
+# Orientations of modules
+
+This file defines orientation of modules.
+
+## Main definitions
+
+* `orientation` is a type class for a choice of orientation of a module.  An orientation may
+be associated with an alternating map or with a basis.
+
+## Implementation notes
+
+`orientation` is defined for an arbitrary index type, but the main intended use case is when
+that index type is a `fintype` and there exists a basis of the same cardinality.
+
+Although `orientation` is a type class, many lemmas have an `orientation` as an explicit
+argument, in order to talk about how two orientations are related or to state results in
+contexts where a particular orientation is not being chosen as canonical.
+
+## References
+
+* https://en.wikipedia.org/wiki/Orientation_(vector_space)
+
+-/
+
+noncomputable theory
+
+section ordered_comm_ring
+
+variables {R : Type*} [ordered_comm_ring R]
+variables {M : Type*} [add_comm_group M] [module R M]
+variables {ι : Type*} [decidable_eq ι]
+
+namespace alternating_map
+
+/-- Two alternating maps have the same orientation if some positive multiples of them are
+equal (in the typical case over a field, this means each is a positive multiple of the other). -/
+def same_orientation (f g : alternating_map R M R ι) : Prop :=
+∃ (r₁ r₂ : R), 0 < r₁ ∧ 0 < r₂ ∧ r₁ • f = r₂ • g
+
+variables (R M ι)
+
+/-- `same_orientation` is symmetric. -/
+lemma symmetric_same_orientation :
+  symmetric (same_orientation : alternating_map R M R ι → alternating_map R M R ι → Prop) :=
+λ f g ⟨r₁, r₂, hr₁, hr₂, h⟩, ⟨r₂, r₁, hr₂, hr₁, h.symm⟩
+
+/-- `same_orientation` is transitive. -/
+lemma transitive_same_orientation :
+  transitive (same_orientation : alternating_map R M R ι → alternating_map R M R ι → Prop) :=
+λ f g h ⟨r₁, r₂, hr₁, hr₂, h₁⟩ ⟨r₃, r₄, hr₃, hr₄, h₂⟩,
+  ⟨r₃ * r₁, r₂ * r₄, mul_pos hr₃ hr₁, mul_pos hr₂ hr₄,
+   by rw [mul_smul, mul_smul, h₁, ←h₂, smul_comm]⟩
+
+section nontrivial
+
+variables [nontrivial R]
+
+/-- `same_orientation` is reflexive. -/
+lemma reflexive_same_orientation :
+  reflexive (same_orientation : alternating_map R M R ι → alternating_map R M R ι → Prop) :=
+λ f, ⟨1, 1, zero_lt_one, zero_lt_one, rfl⟩
+
+/-- `same_orientation` is an equivalence relation. -/
+lemma equivalence_same_orientation :
+  equivalence (same_orientation : alternating_map R M R ι → alternating_map R M R ι → Prop) :=
+⟨reflexive_same_orientation R M ι, symmetric_same_orientation R M ι,
+ transitive_same_orientation R M ι⟩
+
+variables {R M ι}
+
+/-- An alternating map has the same orientation as a positive multiple of itself. -/
+lemma same_orientation_pos_smul_right (f : alternating_map R M R ι) {r : R} (h : 0 < r) :
+  same_orientation f (r • f) :=
```

Comment:
This is true more generally as
```suggestion
  same_orientation f g →same_orientation f (r • g) :=
```
Right? If so, then you could call it `same_orientation.smul_right` for dot notation

Reply:
I've added those dot notation variants (deduced from these ones using transitivity).